tasks/installation.yml

```diff
@@ -1,33 +1,86 @@
 ---
 # tasks file for sap-preconfigure: installation
 
-- name: ensure minimal required packages are installed
+- name: Ensure required package groups are installed
   package:
     state: present
-    name: "{{ sap_preconfigure_min_pkgs }}"
-  when:
-    - sap_preconfigure_min_package_check|bool
-    - not( (sap_preconfigure_min_pkgs is undefined) or (sap_preconfigure_min_pkgs is none) or (sap_preconfigure_min_pkgs | trim == '') )
-
-- name: ensure required package groups are installed
-  package:
-    state: present
-    name: "{{ sap_preconfigure_packagegroups }}"
-  check_mode: "{{ sap_preconfigure_installation_check_only }}"
+    name: "{{ __sap_preconfigure_packagegroups }}"
 
-- name: ensure required packages are installed
+- name: Ensure required packages are installed
   package:
     state: present
-    name: "{{ sap_preconfigure_packages }}"
-  check_mode: "{{ sap_preconfigure_installation_check_only }}"
+    name: "{{ __sap_preconfigure_packages }}"
   register: sap_preconfigure_register_groupinstall
 
-- name: ensure system is updated to the latest patchlevel
+- name: Ensure minimum packages are installed
+  block:
+    - name: check if minimum release needs to be installed
+      shell: |
+             set -x
+             required_pkg={{ pkg | join('-') }}
+             newest=$(echo -e "$required_pkg\n$(rpm -q {{ pkg[0] }} )"| sort -V | tail -1)
+             if [ $newest == $required_pkg ]; then
+                echo $newest
+             fi
+      loop: "{{ __sap_preconfigure_min_pkgs }}"
+      loop_control:
+               loop_var: pkg
+      check_mode: no
+      register: __sap_preconfigure_register_minpkglist
+      changed_when: false
+
+    - name: Initialize an empty list for our strings
+      set_fact:
+        __sap_preconfigure_fact_minpkglist: []
+
+    - name: Create list of packages to be installed
+      set_fact:
+        __sap_preconfigure_fact_minpkglist: "{{ __sap_preconfigure_fact_minpkglist | difference(['']) + [ pkg.stdout ] }}"
+      loop: "{{ __sap_preconfigure_register_minpkglist.results }}"
```

Comment:
This looks odd - in general there is usually a better way to do this rather than Ansible variable = variable + something - but I'm not sure what you are trying to do here

Reply:
We want to make sure that in case a minimum version of a package is required, e.g. setup with version 2.12.2-2.el8_1.1 (see __sap_preconfigure_min_pkgs in vars/*.yml), this package will be installed with the correct minimum version. The goal of "Ensure minimum packages are installed" is to only create a list of packages to be installed if the minimum required version of these packages if necessary.
Can you think of a more compact way to do this?

Reply:
So given a list of packages e.g. `[['nameA','versionA'],['nameB','versionB'],['nameC','versionC']]` - for each package, you want to ensure that the package named `nameA` is installed with version `versionA` or greater?

Reply:
Right, see a similar list of min required package versions in https://github.com/berndfinger/sap-hana-preconfigure/blob/f41334b41de06a92741d09796e9f412537660a45/vars/RedHat_7.yml#L38

Reply:
See also [https://access.redhat.com/solutions/4302501](https://access.redhat.com/solutions/4302501) , which makes me believe that such a feature might be requested/required/useful also outside an SAP environment.

Reply:
The typical way to do this would be to use rpm package dependency management.  E.g. you would have some sap rpm package which has
```
Requires: kernel >= 3.10.....
Requires: setup >= 2.10.2....
```
then, if you did e.g. `yum install sap-pkg`, rpm and yum would automatically ensure that the version of `kernel` was `3.10` or later, `setup` was version `2.10.2` or later.  The problem is that there is no way to do this with `yum` afaik - you cannot do `yum install 'kernel >= 3.10'`.  What I have done in the past is to create "meta" rpm packages - that might be an option for you - dynamically create a rpm spec file from a template, then `rpmbuild` it, then
```
- name: install meta package to ensure dependencies
  package:
    name: /path/to/sap-meta-package.rpm
```
but that might be too much overhead for your use case, brings in rpm devel dependencies, etc.  Anyway, something to consider.

Reply:
Many thanks. Your proposal might even provide some more features which would ease the installation of packages on RHEL systems for SAP. For the time being, I hope the solution provided here, which has worked in the past and is also in use by the current version of sap-hana-preconfigure, is OK to be used in sap-preconfigure as well.

Reply:
ok.  Here is another alternative using jinja2 templating/looping/logic.  I'll leave it up to you to decide if you want to use this or the current way that is proven to work:
```yaml
- name: junk
  hosts: localhost
  tasks:
    - set_fact:
        mypkgs:
          - [ 'kernel' , '5.6.13-201.fc31' ]
          - [ 'glibc' , '2.30-12.fc31' ]
          - [ 'openssl' , '1.1.1g-2.fc31' ]

    - name: Gather package facts
      package_facts:

    - debug:
        msg: facts {{ ansible_facts.packages }}

    - set_fact:
        pkgs_to_install_update: |-
          {% set ns = namespace() %}
          {% set ns.pkgs = [] %}
          {% for mypkg in mypkgs %}
          {%   if not mypkg[0] in ansible_facts.packages %}
          {%     set _ = ns.pkgs.append(mypkg | join('-')) %}
          {%   else %}
          {%     set pkg = ansible_facts.packages[mypkg[0]][-1] %}
          {%     set verrel = pkg.version ~ '-' ~ pkg.release %}
          {%     if mypkg[1] is version(verrel, operator='gt') %}
          {%       set _ = ns.pkgs.append(mypkg | join('-')) %}
          {%     else %}
          {%       set _ = "pkg already installed with same or later version" %}
          {%     endif %}
          {%   endif %}
          {% endfor %}
          {{ ns.pkgs }}

    - debug:
        msg: pkgs_to_install_update {{ pkgs_to_install_update }}
```